marshal/world.go

```diff
@@ -297,6 +296,7 @@ func (m *Marshaler) ClaimPartition(topicName string, partID int) bool {
 // still owning this partition. Returns an error if anything has gone wrong (at which
 // point we can no longer assert we have the lock).
 func (m *Marshaler) Heartbeat(topicName string, partID int, lastOffset int64) error {
+	// cannot be called within topic.lock.RLock() since it calls topic.lock.Lock()
```

Comment:
Was it a pain to debug? I've run into this once or twice, and it's made me think I need to rethink how the locking is done. It's not very idiomatic Go to use locks so much, eh.